ouroboros-network/src/Ouroboros/Network/Protocol/VersionNegotiation/Codec.hs

```diff
@@ -0,0 +1,127 @@
+{-# LANGUAGE BangPatterns        #-}
+{-# LANGUAGE DataKinds           #-}
+{-# LANGUAGE GADTs               #-}
+{-# LANGUAGE KindSignatures      #-}
+{-# LANGUAGE PolyKinds           #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications    #-}
+
+module Ouroboros.Network.Protocol.VersionNegotiation.Codec
+  ( codecVersionNegotiation
+  , SerialiseTerm (..)
+  ) where
+
+import           Control.Monad.Class.MonadST
+import           Control.Monad (unless)
+import           Data.ByteString.Lazy (ByteString)
+import           Data.Map (Map)
+import qualified Data.Map as Map
+
+import qualified Codec.CBOR.Encoding as CBOR
+import qualified Codec.CBOR.Decoding as CBOR
+import qualified Codec.CBOR.Read     as CBOR
+import qualified Codec.CBOR.Term     as CBOR
+import           Codec.Serialise (Serialise)
+import qualified Codec.Serialise     as CBOR
+
+import           Network.TypedProtocol.Codec hiding (encode, decode)
+import           Ouroboros.Network.Codec (mkCodecCborLazyBS)
+
+import           Ouroboros.Network.Protocol.VersionNegotiation.Type
+
+
+-- |
+-- Decoding proposed version is done in two stages.  First decode
+-- a @'CBOR.Term'@, this is done by @'codecVersionNegotiation'@, the second
+-- phase is handled by interface defined in this type class and it is used by
+-- @'versionNegotiationClientPeer'@ and @'versionNeogitationServerPeer'@.
+--
+-- todo: find a better place for this class
+class SerialiseTerm a where
```

Comment:
Is this used?

Reply:
Only in tests, but this is because we don't have any versisions in the library at the moment.  This will change shortly.